src/Passes/docs/src/DeveloperGuide/IntroductionPass.md

```diff
@@ -0,0 +1,190 @@
+  ret i32 %3
+}
+
+define dso_local void @bar(i32 %0, i32 %1) #0 {
+  %3 = alloca i32, align 4
+  %4 = alloca i32, align 4
+  store i32 %0, i32* %3, align 4
+  store i32 %1, i32* %4, align 4
+  %5 = load i32, i32* %3, align 4
+  %6 = load i32, i32* %4, align 4
+  %7 = add nsw i32 %5, %6
+  %8 = call i32 @foo(i32 %7)
+  ret void
+}
+
+define dso_local i32 @main() #0 {
+  %1 = alloca i32, align 4
+  store i32 0, i32* %1, align 4
+  %2 = call i32 @foo(i32 2)
+  call void @bar(i32 3, i32 2)
+  ret i32 0
+}
+```
+
+A stat pass for this code, would collect following statisics:
+
+```text
+Stats for 'foo'
+===========================
+Opcode          # Used
+---------------------------
+load            1
+ret             1
+alloca          1
+store           1
+---------------------------
+
+Stats for 'bar'
+===========================
+Opcode          # Used
+---------------------------
+load            2
+add             1
+ret             1
+alloca          2
+store           2
+call            1
+---------------------------
+
+Stats for 'main'
+===========================
+Opcode          # Used
+---------------------------
+ret             1
+alloca          1
+store           1
+call            2
+---------------------------
+```
+
+**Example 3: Code validation**. A third use case is code validation. For example, one could write a pass to check whether bounds are exceeded on [static arrays][2].
+Note that this is a non-standard usecase as such analysis is usually made using the AST rather than at the IR level.
+
+**References**
+
+- [1] https://github.com/banach-space/llvm-tutor#analysis-vs-transformation-pass
+- [2] https://github.com/victor-fdez/llvm-array-check-pass
+
+## Out-of-source Pass
+
+This library is build as set of out-of-source-passes. All this means is that we will not be downloading the LLVM repository and modifying this repository directly. You can read more [here](https://llvm.org/docs/CMake.html#cmake-out-of-source-pass).
+
+# Library structure for passes
+
+An important part of this PR is that it proposes a structure for passes: It is suggested that each pass has their own subcode base. The reason for this proposal is that it makes it very easy to add and remove passes as well as decide which passes to link against. Each pass is kept in its own subdirectory under `libs`:
+
+```
+Passes
+├── CMakeLists.txt
+└── OpsCounter
+    ├── OpsCounter.cpp
+    └── OpsCounter.hpp
+```
+
+Adding a new pass is easy using the `manage` tool developed in this PR:
+
+```
+./manage create-pass HelloWorld
+Available templates:
+
+1. Function Pass
+
+Select a template:1
+```
+
+which results in a new pass code in the `libs`:
+
+```
+Passes
+├── CMakeLists.txt
+├── HelloWorld
+│   ├── HelloWorld.cpp
+│   ├── HelloWorld.hpp
+│   └── SPECIFICATION.md
+└── OpsCounter
+    ├── OpsCounter.cpp
+    └── OpsCounter.hpp
+```
+
+A full example of how to create a basic function pass is included in the README.md file for anyone interested.
```

Comment:
It may be helpful to link to this README.

Reply:
This file is deprecated.